runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -969,6 +997,271 @@ J9::TransformUtil::canFoldStaticFinalField(TR::Compilation *comp, TR::Node* node
    return TR_maybe;
    }
 
+static bool isTakenSideOfAVirtualGuard(TR::Compilation* comp, TR::Block* block)
+   {
+   // First block can never be taken side
+   if (block == comp->getMethodSymbol()->getFirstTreeTop()->getEnclosingBlock())
+      return false;
+
+   for (auto edge = block->getPredecessors().begin(); edge != block->getPredecessors().end(); ++edge)
+      {
+      TR::Block *pred = toBlock((*edge)->getFrom());
+      TR::Node* predLastRealNode = pred->getLastRealTreeTop()->getNode();
+
+      if (predLastRealNode
+          && predLastRealNode->isTheVirtualGuardForAGuardedInlinedCall()
```

Comment:
This seems fragile - how do you know the call is the last thing in the block? Wouldn't another check be that you want the block to be cold if it is to be considered a virtual guard for guarded inlined call?

Reply:
This is largely recent code already in place but refactored from VP. The caller code of `isTakenSideOfAVirtualGuard` already checks for `block->isCold()` and skip folding if cold. So this `isTakenSideOfAVirtualGuard` helper seems to perform some extra diligence in hopes to catch some more cases. Given that we already skips folding for cold should this virtual guard check be removed since redundant? @andrewcraik

Reply:
We still need this check because taken side of a virtual guard is not always cold - consider a profiled guard for a method called 60% of the time - the taken side of the guard will be given 40% of the block frequency and will remain non-cold to indicate we expect it to run.

Given the code is already in use we can leave this for now - we'll address it if it becomes a problem.

Reply:
@andrewcraik `isTheVirtualGuardForAGuardedInlinedCall` checks if a node is a guard. A guard is the last thing in a block.
We want to skip folding on the taken side because the taken side is just a call and we won't benefit from the folding.